In [198]:
import pandas as pd
from pandasql import sqldf
import numpy as np
from geopy.geocoders import Nominatim
from geopy.distance import geodesic
import warnings

In [199]:
warnings.filterwarnings("ignore")

In [200]:
data = pd.read_csv("FlightDistanceTest.csv")

Checking the data information, i.e. the data types, column names, and the number of null values.

In [201]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Normalised City Pair  3000 non-null   object 
 1   Departure Code        3000 non-null   object 
 2   Arrival Code          3000 non-null   object 
 3   Departure_lat         3000 non-null   float64
 4   Departure_lon         3000 non-null   float64
 5   Arrival_lat           3000 non-null   float64
 6   Arrival_lon           3000 non-null   float64
dtypes: float64(4), object(3)
memory usage: 164.2+ KB


In [202]:
data.isnull().sum()

Normalised City Pair    0
Departure Code          0
Arrival Code            0
Departure_lat           0
Departure_lon           0
Arrival_lat             0
Arrival_lon             0
dtype: int64

In [203]:
data.head(10)

Normalised City Pair Departure Code  \
0  London, United Kingdom - New York, United States Of America            LHR   
1          Johannesburg, South Africa - London, United Kingdom            JNB   
2  London, United Kingdom - New York, United States Of America            LHR   
3          Johannesburg, South Africa - London, United Kingdom            JNB   
4                London, United Kingdom - Singapore, Singapore            SIN   
5  London, United Kingdom - New York, United States Of America            JFK   
6  London, United Kingdom - New York, United States Of America            LCY   
7    London, United Kingdom - Newark, United States Of America            EWR   
8                       Bombay, India - London, United Kingdom            LHR   
9         Dubai, United Arab Emirates - London, United Kingdom            DXB   

  Arrival Code  Departure_lat  Departure_lon  Arrival_lat  Arrival_lon  
0          JFK           51.5          -0.45        40.64       -73.79  
1          LHR          -26.1          28.23        51.47        -0.45  
2          JFK           51.5          -0.45        40.64       -73.79  
3          LHR          -26.1          28.23        51.47        -0.45  
4          LHR            1.3         103.98        51.47        -0.45  
5          LCY           40.6         -73.79        51.50         0.05  
6          JFK           51.5           0.05        40.64       -73.79  
7          LHR           40.7         -74.18        51.47        -0.45  
8          BOM           51.5          -0.45        19.09        72.86  
9          LHR           25.3          55.35        51.47        -0.45

Calculating route distances in miles based on latitude and longitude.

In [204]:
pd.set_option("display.max_colwidth", 300)

At this point, my intention was to split the strings from the "Normalized City Pair" column to create new columns for the city of departure, city of arrival, country of departure, and country of arrival. The goal was to verify the exact addresses for each coordinate and to slice strings in the "Normalized City Pair" column to check if any of the words matched the generated address. After using the *geopy* library to find the exact addresses of the coordinates, I quickly realized that this approach was impractical. This was because some addresses use different character sets (e.g. Arabic letters) or the exact address does not include the city name. For example, the exact address for the coordinates "-26.1, 28.23" is "Kosmos Road, Ekurhuleni Ward 16, Kempton Park, City of Ekurhuleni Metropolitan Municipality, Gauteng, 1627, South Africa." The corresponding row in the "Normalized City Pair" column for that location is "Johannesburg", which my method would not have detected since "Johannesburg" is not mentioned in the exact address *geopy* provided.

In [205]:
def location(latitude_and_longitude: str) -> str: # returns location (exact address) of the input coordinates
    geolocator = Nominatim(user_agent="locator")
    geolocation = geolocator.reverse(latitude_and_longitude)
    return geolocation.address if geolocation else ""
print(location("-26.1, 28.23")) # example

Kosmos Road, Ekurhuleni Ward 16, Kempton Park, City of Ekurhuleni Metropolitan Municipality, Gauteng, 1627, South Africa


Since "Normalizsed City Pair" column has been removed from the scope of my analysis, another potential anchor point could be to verify if airport codes are consistent for each set of geographic coordinates. First, I will create a temporary table that will store concatenated values of latitude and longitude (for easier interpretation), along with the airport codes for both departure and arrival. This will enable me to search for data inconsistencies using SQL.

In [206]:
# concatenating longitude and latitude into a single column
data["departure_coords"] = data["Departure_lat"].astype(str) + ", " + data["Departure_lon"].astype(str)
data["arrival_coords"] = data["Arrival_lat"].astype(str) + ", " + data["Arrival_lon"].astype(str)

In [207]:
departure_code = pd.DataFrame(data["Departure Code"].copy())
arrival_code = pd.DataFrame(data["Arrival Code"].copy())

temp_df = departure_code.join(data["departure_coords"], how = "inner")
temp_df = temp_df.join(arrival_code)
temp_df = temp_df.join(data["arrival_coords"], how = "inner")
temp_df.columns = ["departure_code", "departure_coords", "arrival_code", "arrival_coords"] # renaming columns for easier handling in sql

In [208]:
temp_df

departure_code departure_coords arrival_code arrival_coords
0               LHR      51.5, -0.45          JFK  40.64, -73.79
1               JNB     -26.1, 28.23          LHR   51.47, -0.45
2               LHR      51.5, -0.45          JFK  40.64, -73.79
3               JNB     -26.1, 28.23          LHR   51.47, -0.45
4               SIN      1.3, 103.98          LHR   51.47, -0.45
...             ...              ...          ...            ...
2995            LGA     40.8, -73.88          YYZ  43.68, -79.61
2996            EGE     41.3, -85.28          EWR  40.69, -74.18
2997            SOU      51.0, -1.36          IOM   54.09, -4.63
2998            YYZ     43.7, -79.61          YOW  45.32, -75.67
2999            HEL      60.3, 24.97          FRA    50.05, 8.56

[3000 rows x 4 columns]

Now, I will proceed with querying the above table to determine if there are any coordinates that are associated with more than one airport code (which of course, shouldn't occur).

In [209]:
sqldf(""" 
    SELECT departure_coords, COUNT(DISTINCT departure_code) AS count_of_distinct_departure_codes
    FROM temp_df
    GROUP BY departure_coords
    HAVING COUNT(DISTINCT departure_code) > 1
    ORDER BY count_of_distinct_departure_codes DESC
""")

departure_coords  count_of_distinct_departure_codes
0       51.5, -0.45                                 39
1        50.1, 8.56                                 26
2       25.3, 55.35                                 18
3        49.0, 2.54                                 14
4      41.6, -98.93                                 14
5     34.1, -118.24                                 11
6        47.5, 8.57                                 10
7       1.3, 103.98                                 10
8        52.3, 4.75                                  9
9       48.3, 11.77                                  9
10     42.9, -83.53                                  9
11     40.6, -73.79                                  9
12    33.4, -110.87                                  9
13      25.3, 51.56                                  9
14     -26.1, 28.23                                  9
15     32.9, -97.04                                  8
16     40.7, -74.18                                  7
17     35.8, 140.39                                  7
18     30.0, -95.34                                  7
19     22.3, 113.94                                  7
20      40.5, -3.58                                  6
21      48.2, 16.37                                  5
22      44.9, -93.2                                  5
23      41.0, 28.99                                  5
24    39.8, -104.86                                  5
25     39.0, -77.45                                  5
26    37.3, -107.75                                  5
27     35.2, -80.95                                  5
28      -1.3, 36.81                                  5
29      55.6, 12.64                                  4
30     45.1, -89.01                                  4
31     43.7, -79.61                                  4
32    40.8, -111.89                                  4
33    37.6, -122.39                                  4
34     33.1, -88.31                                  4
35      28.7, 77.26                                  4
36       51.5, 0.05                                  3
37    49.2, -123.17                                  3
38       45.5, 9.18                                  3
39       41.9, 12.5                                  3
40      38.8, -9.13                                  3
41     13.7, 100.77                                  3
42    -23.4, -46.48                                  3
43      60.3, 24.97                                  2
44        53.4, 0.1                                  2
45      52.5, -1.73                                  2
46      51.2, -0.16                                  2
47      51.0, -1.36                                  2
48     39.9, -75.59                                  2
49     37.5, 126.66                                  2
50    33.4, -112.03                                  2
51      23.6, 58.29                                  2
52    21.3, -157.92                                  2
53    -37.7, 144.85                                  2
54    -33.9, 151.23                                  2
55     -20.4, 57.68                                  2
56     -20.0, 23.43                                  2
57    -12.0, -77.11                                  2

In [210]:
sqldf("""
    SELECT 'asssuming_1_correct' AS type, SUM(count_of_distinct_departure_codes) -  COUNT(departure_coords) as unique_incorrect_airport_codes
    FROM (
        SELECT departure_coords, COUNT(DISTINCT departure_code) AS count_of_distinct_departure_codes
        FROM temp_df
        GROUP BY departure_coords
        HAVING COUNT(DISTINCT departure_code) > 1
    ) temp
UNION ALL
    SELECT 'asssuming_none_correct' AS type, SUM(count_of_distinct_departure_codes) as unique_incorrect_airport_codes
    FROM (
        SELECT departure_coords, COUNT(DISTINCT departure_code) AS count_of_distinct_departure_codes
        FROM temp_df
        GROUP BY departure_coords
        HAVING COUNT(DISTINCT departure_code) > 1
    ) temp
""")

type  unique_incorrect_airport_codes
0     asssuming_1_correct                             308
1  asssuming_none_correct                             366

As we can see, some coordinates are associated with multiple airport codes, each with at least two codes assigned. It is safe to assume that the number of unique, incorrectly assigned departure airport codes varies between 308 and 366, depending on whether we assume at least one or none of the codes per set of coordinates is correct. The coordinates '51.5, -0.45' have the largest number of airport codes assigned to them. Let's proceed and check the specific codes matched with these coordinates.

In [211]:
sqldf("""
    SELECT DISTINCT departure_code
    FROM temp_df
    WHERE departure_coords = '51.5, -0.45'
    ORDER BY departure_code  
""")

departure_code
0             AMS
1             ARN
2             BOM
3             BRU
4             CDG
5             CPT
6             DEL
7             DFW
8             DOH
9             DXB
10            EDI
11            EWR
12            FRA
13            GIG
14            GLA
15            GRU
16            GVA
17            HEL
18            ICN
19            JFK
20            LHR
21            LIN
22            LIS
23            LON
24            LOS
25            MAA
26            MAD
27            MAN
28            MIA
29            NCL
30            NRT
31            NYC
32            ORD
33            OSL
34            PHL
35            PRG
36            SFO
37            SIN
38            ZRH

There are **39 unique** airport codes associated with the coordinates "51.5, -0.45". We can use the location function mentioned earlier to decode these coordinates and determine the correct code that should be assigned to them. To verify the airport code we can look for "airport codes list" on Google for example, and use any of the resulting website.

In [212]:
location("51.5 -0.45")

'Stockley Road, West Drayton, London Borough of Hillingdon, London, Greater London, England, UB7 9BN, United Kingdom'

The coordinates "51.5, -0.45" correspond to London. Therefore, any airport codes other than "LHR" associated with these specific coordinates are, in fact, errors in the dataset.

We can also display all of the airport codes assigned to the previously mentioned coordinates.

In [213]:
sqldf("""
    SELECT departure_coords, GROUP_CONCAT(DISTINCT departure_code) AS departure_codes
    FROM (
        SELECT DISTINCT departure_coords, departure_code
        FROM temp_df
    ) unique_codes
    GROUP BY departure_coords
    HAVING COUNT(DISTINCT departure_code) > 1
""")

departure_coords  \
0       -1.3, 36.81   
1     -12.0, -77.11   
2      -20.0, 23.43   
3      -20.4, 57.68   
4     -23.4, -46.48   
5      -26.1, 28.23   
6     -33.9, 151.23   
7     -37.7, 144.85   
8       1.3, 103.98   
9      13.7, 100.77   
10    21.3, -157.92   
11     22.3, 113.94   
12      23.6, 58.29   
13      25.3, 51.56   
14      25.3, 55.35   
15      28.7, 77.26   
16     30.0, -95.34   
17     32.9, -97.04   
18     33.1, -88.31   
19    33.4, -110.87   
20    33.4, -112.03   
21    34.1, -118.24   
22     35.2, -80.95   
23     35.8, 140.39   
24    37.3, -107.75   
25     37.5, 126.66   
26    37.6, -122.39   
27      38.8, -9.13   
28     39.0, -77.45   
29    39.8, -104.86   
30     39.9, -75.59   
31      40.5, -3.58   
32     40.6, -73.79   
33     40.7, -74.18   
34    40.8, -111.89   
35      41.0, 28.99   
36     41.6, -98.93   
37       41.9, 12.5   
38     42.9, -83.53   
39     43.7, -79.61   
40      44.9, -93.2   
41     45.1, -89.01   
42       45.5, 9.18   
43       47.5, 8.57   
44      48.2, 16.37   
45      48.3, 11.77   
46       49.0, 2.54   
47    49.2, -123.17   
48       50.1, 8.56   
49      51.0, -1.36   
50      51.2, -0.16   
51      51.5, -0.45   
52       51.5, 0.05   
53       52.3, 4.75   
54      52.5, -1.73   
55        53.4, 0.1   
56      55.6, 12.64   
57      60.3, 24.97   

                                                                                                                                                departure_codes  
0                                                                                                                                           NBO,LHR,EBB,JNB,ACC  
1                                                                                                                                                       LIM,EZE  
2                                                                                                                                                       MUB,GBE  
3                                                                                                                                                       MRU,SEZ  
4                                                                                                                                                   GRU,LHR,JFK  
5                                                                                                                           JNB,CAI,GBE,DXB,HKG,DUR,GRU,LHR,DOH  
6                                                                                                                                                       SYD,NRT  
7                                                                                                                                                       MEL,LHR  
8                                                                                                                       SIN,LHR,MNL,EWR,CGK,AKL,NRT,JFK,KUL,FRA  
9                                                                                                                                                   BKK,HKG,FRA  
10                                                                                                                                                      HNL,KOA  
11                                                                                                                                  HKG,LHR,TPE,MEL,PER,JFK,SGN  
12                                                                                                                                                      MCT,DXB  
13                                                                                                                          DOH,AUH,SEZ,GVA,BLR,LHR,SIN,MAA,DXB  
14                                                                                      DXB,BOM,MAN,JNB,SIN,MAA,BHX,LGW,LHR,NCL,KHI,MNL,JFK,DOH,DEL,ACC,PRG,FCO  
15                                                   

Now, I will query the columns containing information about arrival coordinates and arrival airport codes.

In [214]:
sqldf(""" 
    SELECT arrival_coords, COUNT(DISTINCT arrival_code) AS count_distinct_arrival_codes
    FROM temp_df
    GROUP BY arrival_coords
    HAVING COUNT(DISTINCT arrival_code) > 1;
""")

arrival_coords  count_distinct_arrival_codes
0  40.07, 116.58                             2
1    45.47, 9.18                             2
2    49.01, 2.54                             2
3   51.47, -0.45                             2

In [215]:
sqldf("""
    SELECT arrival_coords,
    GROUP_CONCAT(DISTINCT arrival_code) as arrival_codes
    FROM (
        SELECT DISTINCT arrival_coords, arrival_code
        FROM temp_df
    ) unique_codes
    GROUP BY arrival_coords
    HAVING COUNT(DISTINCT arrival_code) > 1
""")

arrival_coords arrival_codes
0  40.07, 116.58       PEK,BJS
1    45.47, 9.18       LIN,MXP
2    49.01, 2.54       CDG,PAR
3   51.47, -0.45       LHR,LON

There are significantly fewer errors in the arrival coordinates and codes. However, some errors still exist.

Now, I will calculate the distances between provided coordinates and compare them with distances calculated based on airport codes and coordinates using this [database](https://raw.githubusercontent.com/jpatokal/openflights/master/data/airports.dat).

In [216]:
airports = pd.read_csv("https://raw.githubusercontent.com/jpatokal/openflights/master/data/airports.dat", header = None)

In [217]:
airports

0                                            1             2   \
0         1                               Goroka Airport        Goroka   
1         2                               Madang Airport        Madang   
2         3                 Mount Hagen Kagamuga Airport   Mount Hagen   
3         4                               Nadzab Airport        Nadzab   
4         5  Port Moresby Jacksons International Airport  Port Moresby   
...     ...                                          ...           ...   
7693  14106                          Rogachyovo Air Base        Belaya   
7694  14107                        Ulan-Ude East Airport      Ulan Ude   
7695  14108                         Krechevitsy Air Base      Novgorod   
7696  14109                  Desierto de Atacama Airport       Copiapo   
7697  14110                           Melitopol Air Base     Melitopol   

                    3    4     5          6           7     8   9   10  \
0     Papua New Guinea  GKA  AYGA  -6.081690  145.391998  5282  10   U   
1     Papua New Guinea  MAG  AYMD  -5.207080  145.789001    20  10   U   
2     Papua New Guinea  HGU  AYMH  -5.826790  144.296005  5388  10   U   
3     Papua New Guinea  LAE  AYNZ  -6.569803  146.725977   239  10   U   
4     Papua New Guinea  POM  AYPY  -9.443380  147.220001   146  10   U   
...                ...  ...   ...        ...         ...   ...  ..  ..   
7693            Russia   \N  ULDA  71.616699   52.478298   272  \N  \N   
7694            Russia   \N  XIUW  51.849998  107.737999  1670  \N  \N   
7695            Russia   \N  ULLK  58.625000   31.385000    85  \N  \N   
7696             Chile  CPO  SCAT -27.261200  -70.779198   670  \N  \N   
7697           Ukraine   \N  UKDM  46.880001   35.305000     0  \N  \N   

                        11       12           13  
0     Pacific/Port_Moresby  airport  OurAirports  
1     Pacific/Port_Moresby  airport  OurAirports  
2     Pacific/Port_Moresby  airport  OurAirports  
3     Pacific/Port_Moresby  airport  OurAirports  
4     Pacific/Port_Moresby  airport  OurAirports  
...                    ...      ...          ...  
7693                    \N  airport  OurAirports  
7694                    \N  airport  OurAirports  
7695                    \N  airport  OurAirports  
7696                    \N  airport  OurAirports  
7697                    \N  airport  OurAirports  

[7698 rows x 14 columns]

In [218]:
drop = [0, 1, 3, 5, 8, 9, 10, 11, 12, 13] # dropping useless columns
airports = airports.drop(drop, axis = "columns")

In [219]:
airports.columns = ["city", "airport_code", "latitude", "longitude"]

In [220]:
airports["latitude"] = airports["latitude"].round(2)
airports["longitude"] = airports["longitude"].round(2)

In [221]:
airports["coordinates"] = airports["latitude"].astype(str) + ", " + airports["longitude"].astype(str) # geopy handles strings

In [222]:
drop = ["latitude", "longitude"]
airports = airports.drop(drop, axis = "columns")
airports

city airport_code     coordinates
0           Goroka          GKA   -6.08, 145.39
1           Madang          MAG   -5.21, 145.79
2      Mount Hagen          HGU    -5.83, 144.3
3           Nadzab          LAE   -6.57, 146.73
4     Port Moresby          POM   -9.44, 147.22
...            ...          ...             ...
7693        Belaya           \N    71.62, 52.48
7694      Ulan Ude           \N   51.85, 107.74
7695      Novgorod           \N    58.62, 31.39
7696       Copiapo          CPO  -27.26, -70.78
7697     Melitopol           \N     46.88, 35.3

[7698 rows x 3 columns]

In [223]:
data

Normalised City Pair  \
0                 London, United Kingdom - New York, United States Of America   
1                         Johannesburg, South Africa - London, United Kingdom   
2                 London, United Kingdom - New York, United States Of America   
3                         Johannesburg, South Africa - London, United Kingdom   
4                               London, United Kingdom - Singapore, Singapore   
...                                                                       ...   
2995                     New York, United States Of America - Toronto, Canada   
2996  Newark, United States Of America - Vail/eagle, United States Of America   
2997                Isle Of Man, United Kingdom - Southampton, United Kingdom   
2998                                         Ottawa, Canada - Toronto, Canada   
2999                                   Frankfurt, Germany - Helsinki, Finland   

     Departure Code Arrival Code  Departure_lat  Departure_lon  Arrival_lat  \
0               LHR          JFK           51.5          -0.45        40.64   
1               JNB          LHR          -26.1          28.23        51.47   
2               LHR          JFK           51.5          -0.45        40.64   
3               JNB          LHR          -26.1          28.23        51.47   
4               SIN          LHR            1.3         103.98        51.47   
...             ...          ...            ...            ...          ...   
2995            LGA          YYZ           40.8         -73.88        43.68   
2996            EGE          EWR           41.3         -85.28        40.69   
2997            SOU          IOM           51.0          -1.36        54.09   
2998            YYZ          YOW           43.7         -79.61        45.32   
2999            HEL          FRA           60.3          24.97        50.05   

      Arrival_lon departure_coords arrival_coords  
0          -73.79      51.5, -0.45  40.64, -73.79  
1           -0.45     -26.1, 28.23   51.47, -0.45  
2          -73.79      51.5, -0.45  40.64, -73.79  
3           -0.45     -26.1, 28.23   51.47, -0.45  
4           -0.45      1.3, 103.98   51.47, -0.45  
...           ...              ...            ...  
2995       -79.61     40.8, -73.88  43.68, -79.61  
2996       -74.18     41.3, -85.28  40.69, -74.18  
2997        -4.63      51.0, -1.36   54.09, -4.63  
2998       -75.67     43.7, -79.61  45.32, -75.67  
2999         8.56      60.3, 24.97    50.05, 8.56  

[3000 rows x 9 columns]

In [224]:
drop = ["Departure_lat", "Departure_lon", "Arrival_lat", "Arrival_lon"]
data = data.drop(drop, axis = "columns")
data

Normalised City Pair  \
0                 London, United Kingdom - New York, United States Of America   
1                         Johannesburg, South Africa - London, United Kingdom   
2                 London, United Kingdom - New York, United States Of America   
3                         Johannesburg, South Africa - London, United Kingdom   
4                               London, United Kingdom - Singapore, Singapore   
...                                                                       ...   
2995                     New York, United States Of America - Toronto, Canada   
2996  Newark, United States Of America - Vail/eagle, United States Of America   
2997                Isle Of Man, United Kingdom - Southampton, United Kingdom   
2998                                         Ottawa, Canada - Toronto, Canada   
2999                                   Frankfurt, Germany - Helsinki, Finland   

     Departure Code Arrival Code departure_coords arrival_coords  
0               LHR          JFK      51.5, -0.45  40.64, -73.79  
1               JNB          LHR     -26.1, 28.23   51.47, -0.45  
2               LHR          JFK      51.5, -0.45  40.64, -73.79  
3               JNB          LHR     -26.1, 28.23   51.47, -0.45  
4               SIN          LHR      1.3, 103.98   51.47, -0.45  
...             ...          ...              ...            ...  
2995            LGA          YYZ     40.8, -73.88  43.68, -79.61  
2996            EGE          EWR     41.3, -85.28  40.69, -74.18  
2997            SOU          IOM      51.0, -1.36   54.09, -4.63  
2998            YYZ          YOW     43.7, -79.61  45.32, -75.67  
2999            HEL          FRA      60.3, 24.97    50.05, 8.56  

[3000 rows x 5 columns]

In [225]:
data = data.rename(columns = {"departure_coords": "dep_coords_ORIGINAL", "arrival_coords": "arr_coords_ORIGINAL"})

Checking unique airport codes in the dataset

In [226]:
codes = set(data['Departure Code'].unique()).union(set(data['Arrival Code'].unique()))

Dropping airport codes that are not present in original dataset

In [227]:
airports = airports[airports['airport_code'].isin(codes)]

In [228]:
airports.reset_index(drop = True)

city airport_code     coordinates
0              Keflavik          KEF   63.99, -22.61
1              Edmonton          YEG  53.31, -113.58
2               Halifax          YHZ   44.88, -63.51
3                Ottawa          YOW   45.32, -75.67
4                Quebec          YQB   46.79, -71.39
..                  ...          ...             ...
391  Hilton Head Island          HHH    32.22, -80.7
392                Doha          DOH    25.27, 51.61
393           Hyderabad          HYD    17.23, 78.43
394            Istanbul          IST    41.28, 28.75
395           Islamabad          ISB    33.56, 72.85

[396 rows x 3 columns]

In [229]:
data

Normalised City Pair  \
0                 London, United Kingdom - New York, United States Of America   
1                         Johannesburg, South Africa - London, United Kingdom   
2                 London, United Kingdom - New York, United States Of America   
3                         Johannesburg, South Africa - London, United Kingdom   
4                               London, United Kingdom - Singapore, Singapore   
...                                                                       ...   
2995                     New York, United States Of America - Toronto, Canada   
2996  Newark, United States Of America - Vail/eagle, United States Of America   
2997                Isle Of Man, United Kingdom - Southampton, United Kingdom   
2998                                         Ottawa, Canada - Toronto, Canada   
2999                                   Frankfurt, Germany - Helsinki, Finland   

     Departure Code Arrival Code dep_coords_ORIGINAL arr_coords_ORIGINAL  
0               LHR          JFK         51.5, -0.45       40.64, -73.79  
1               JNB          LHR        -26.1, 28.23        51.47, -0.45  
2               LHR          JFK         51.5, -0.45       40.64, -73.79  
3               JNB          LHR        -26.1, 28.23        51.47, -0.45  
4               SIN          LHR         1.3, 103.98        51.47, -0.45  
...             ...          ...                 ...                 ...  
2995            LGA          YYZ        40.8, -73.88       43.68, -79.61  
2996            EGE          EWR        41.3, -85.28       40.69, -74.18  
2997            SOU          IOM         51.0, -1.36        54.09, -4.63  
2998            YYZ          YOW        43.7, -79.61       45.32, -75.67  
2999            HEL          FRA         60.3, 24.97         50.05, 8.56  

[3000 rows x 5 columns]

In [230]:
departure_merged = pd.merge(data,
                            airports[["airport_code", "coordinates"]],
                            left_on = "Departure Code",
                            right_on = "airport_code",
                            how = "left").drop("airport_code", axis = "columns")

departure_merged.rename(columns = {"coordinates": "dep_coords"}, inplace = True)


merged = pd.merge(departure_merged,
                        airports[["airport_code", "coordinates"]],
                        left_on = "Arrival Code",
                        right_on = "airport_code",
                        how = "left").drop("airport_code", axis = "columns")

merged.rename(columns = {"coordinates": "arr_coords"}, inplace = True)
merged

Normalised City Pair  \
0                 London, United Kingdom - New York, United States Of America   
1                         Johannesburg, South Africa - London, United Kingdom   
2                 London, United Kingdom - New York, United States Of America   
3                         Johannesburg, South Africa - London, United Kingdom   
4                               London, United Kingdom - Singapore, Singapore   
...                                                                       ...   
2995                     New York, United States Of America - Toronto, Canada   
2996  Newark, United States Of America - Vail/eagle, United States Of America   
2997                Isle Of Man, United Kingdom - Southampton, United Kingdom   
2998                                         Ottawa, Canada - Toronto, Canada   
2999                                   Frankfurt, Germany - Helsinki, Finland   

     Departure Code Arrival Code dep_coords_ORIGINAL arr_coords_ORIGINAL  \
0               LHR          JFK         51.5, -0.45       40.64, -73.79   
1               JNB          LHR        -26.1, 28.23        51.47, -0.45   
2               LHR          JFK         51.5, -0.45       40.64, -73.79   
3               JNB          LHR        -26.1, 28.23        51.47, -0.45   
4               SIN          LHR         1.3, 103.98        51.47, -0.45   
...             ...          ...                 ...                 ...   
2995            LGA          YYZ        40.8, -73.88       43.68, -79.61   
2996            EGE          EWR        41.3, -85.28       40.69, -74.18   
2997            SOU          IOM         51.0, -1.36        54.09, -4.63   
2998            YYZ          YOW        43.7, -79.61       45.32, -75.67   
2999            HEL          FRA         60.3, 24.97         50.05, 8.56   

          dep_coords     arr_coords  
0       51.47, -0.46  40.64, -73.78  
1      -26.14, 28.25   51.47, -0.46  
2       51.47, -0.46  40.64, -73.78  
3      -26.14, 28.25   51.47, -0.46  
4       1.35, 103.99   51.47, -0.46  
...              ...            ...  
2995   40.78, -73.87  43.68, -79.63  
2996  39.64, -106.92  40.69, -74.17  
2997    50.95, -1.36   54.08, -4.62  
2998   43.68, -79.63  45.32, -75.67  
2999    60.32, 24.96    50.03, 8.57  

[3000 rows x 7 columns]

Changing order of columns

In [231]:
new_order = ["Normalised City Pair", "Departure Code", "dep_coords",
             "dep_coords_ORIGINAL", "Arrival Code", "arr_coords", "arr_coords_ORIGINAL"]
data = merged[new_order]

In [232]:
data.isnull().sum()

Normalised City Pair     0
Departure Code           0
dep_coords               8
dep_coords_ORIGINAL      0
Arrival Code             0
arr_coords              13
arr_coords_ORIGINAL      0
dtype: int64

Imported database couldn't find matching coordinates for provided airport codes, resulting in a few null values after joining the tables. I will ignore these for now as they can be handled manually.

In [233]:
def distance(row, departure, arrival):
    departure = row[departure]
    arrival = row[arrival]

    if pd.isna(departure) or pd.isna(arrival): # handling missing data
        return None

    distances = geodesic(departure, arrival).miles
    return round(distances, 2)

In [234]:
data["distance_in_miles_calculated_with_original_coords"] = data.apply(distance, departure = "dep_coords_ORIGINAL", arrival = "arr_coords_ORIGINAL", axis = "columns")
data["distance_in_miles_calculated__between_airport_codes"] = data.apply(distance, departure = "dep_coords", arrival = "arr_coords", axis = "columns")

In [235]:
data["distance_diff"] = abs(data["distance_in_miles_calculated_with_original_coords"] - data["distance_in_miles_calculated__between_airport_codes"]) # calculating absolute diff

In [236]:
data = data.sort_values(by = "distance_diff", ascending = False)

In [237]:
data.to_excel("result.xlsx", index = False)

### Useful Links and References:
* [Getting distance between two points based on latitude/longitude](https://stackoverflow.com/questions/19412462/getting-distance-between-two-points-based-on-latitude-longitude)
* [Airport codes](https://www.ccra.com/airport-codes/)
* [Get address from given coordinate using python](https://stackoverflow.com/questions/60928516/get-address-from-given-coordinate-using-python)
* [How to Use SQL in pandas Using pandasql Queries](https://www.datacamp.com/tutorial/how-to-use-sql-in-pandas-using-pandasql-queries)
* [Using an API to calculate distance between two airports](https://stackoverflow.com/questions/37572731/using-an-api-to-calculate-distance-between-two-airports-two-columns-within-r)
* [Database of airports](https://raw.githubusercontent.com/jpatokal/openflights/master/data/airports.dat)